cornac/models/lightgcn2/recom_lightgcn.py

```diff
@@ -0,0 +1,280 @@
+
+        val_set: :obj:`cornac.data.Dataset`, optional, default: None
+            User-Item preference data for model selection purposes (e.g., early stopping).
+
+        Returns
+        -------
+        self : object
+        """
+        Recommender.fit(self, train_set, val_set)
+
+        if not self.trainable:
+            return self
+
+        # model setup
+        import torch
+        from .lightgcn import Model
+        from .lightgcn import construct_graph
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        if self.seed is not None:
+            torch.manual_seed(self.seed)
+            if torch.cuda.is_available():
+                torch.cuda.manual_seed_all(self.seed)
+
+        graph = construct_graph(train_set).to(self.device)
+        model = Model(
+            graph,
+            self.emb_size,
+            self.num_layers,
+            self.lambda_reg,
+        ).to(self.device)
+
+        optimizer = torch.optim.Adam(model.parameters(), lr=self.learning_rate)
+
+        # model training
+        pbar = trange(
+            self.num_epochs,
+            desc="Training",
+            unit="iter",
+            position=0,
+            leave=False,
+            disable=not self.verbose,
+        )
+        for _ in pbar:
+            model.train()
+            accum_loss = 0.0
+            for batch_u, batch_i, batch_j in tqdm(
+                train_set.uij_iter(
+                    batch_size=self.train_batch_size,
+                    shuffle=True,
+                ),
+                desc="Epoch",
+                total=train_set.num_batches(self.train_batch_size),
+                leave=False,
+                position=1,
+                disable=not self.verbose,
+            ):
+                u_g_embeddings, pos_i_g_embeddings, neg_i_g_embeddings = model(
+                    graph, batch_u, batch_i, batch_j
+                )
+
+                batch_loss, batch_bpr_loss, batch_reg_loss = model.loss_fn(
+                    u_g_embeddings, pos_i_g_embeddings, neg_i_g_embeddings
+                )
+                accum_loss += batch_loss.cpu().item() * len(batch_u)
+
+                optimizer.zero_grad()
+                batch_loss.backward()
+                optimizer.step()
+
+            accum_loss /= len(train_set.uir_tuple[0])  # normalize over all observations
+            pbar.set_postfix(loss=accum_loss)
+
+            # store user and item embedding matrices for prediction
+            model.eval()
+            u_embs, i_embs, _ = model(graph)
+            # we will use numpy for faster prediction in the score function, no need torch
+            self.U = u_embs.cpu().detach().numpy()
+            self.V = i_embs.cpu().detach().numpy()
+
+            if self.early_stopping is not None and self.early_stop(
+                **self.early_stopping
+            ):
+                break
+
+    def monitor_value(self):
```

Comment:
I think we'll be using recall@20 to do early stopping similar to NGCF instead of the loss value